File: packages/node/src/providers/microsoft/token-validator.ts
import { decodeJwt, errors as joseErrors, jwtVerify, type JWTPayload } from "jose";
import { type RawClaims } from "../../core/claims.js";
import { type NormalizedMicrosoftConfig, type TokenVersion } from "../../config/types.js";
import {
  AuthenticationError,
  InvalidAudienceError,
  InvalidIssuerError,
  InvalidSignatureError,
  InvalidTokenError,
  TokenExpiredError,
  TokenNotYetValidError,
  type AuthError,
} from "../../errors/index.js";
import { type MicrosoftJwksClient } from "./jwks-client.js";
import { assertTenantAllowed, resolveExpectedIssuer } from "./issuer.js";

/**
 * Only RS256 is accepted. Pinning the algorithm set is a hard security boundary:
 * it blocks `alg: none` and RSA/HMAC confusion attacks. This is intentionally not
 * configurable — Entra signs with RS256.
 */
const ACCEPTED_ALGORITHMS = ["RS256"] as const;

/**
 * Validates Microsoft Entra ID tokens end-to-end: signature, issuer, audience,
 * expiry (`exp`), not-before (`nbf`), tenant, and token version.
 *
 * Flow:
 *  1. Decode (unverified) just enough to read `tid` and `ver` so the correct
 *     issuer can be computed — `common`/`organizations` have no fixed issuer.
 *  2. Cryptographically verify signature + standard claims against the resolved
 *     issuer and configured audience(s) via jose.
 *  3. Assert tenant policy and token-version policy on the *verified* payload.
 *
 * No claim is trusted until after step 2.
 */
export class MicrosoftTokenValidator {
  public constructor(
    private readonly config: NormalizedMicrosoftConfig,
    private readonly jwksClient: MicrosoftJwksClient,
  ) {}

  public async validate(token: string): Promise<RawClaims> {
    // --- Step 1: peek (untrusted) to resolve issuer ------------------------
    const peeked = this.decode(token);
    const version = this.resolveVersion(peeked);
    const tid = this.requireTid(peeked);
    const expectedIssuer = resolveExpectedIssuer(tid, version, this.config);

    // --- Step 2: cryptographic verification --------------------------------
    const payload = await this.verifySignatureAndClaims(token, expectedIssuer);

    // --- Step 3: policy on the verified payload ----------------------------
    const verifiedTid = this.requireTid(payload);
    assertTenantAllowed(verifiedTid, this.config);
    this.assertVersionAccepted(this.resolveVersion(payload));

    return payload;
  }

  // --- internals -----------------------------------------------------------

  private decode(token: string): JWTPayload {
    try {
      return decodeJwt(token);
    } catch (cause) {
      throw new InvalidTokenError("Token could not be decoded", {
        cause: cause instanceof Error ? cause.message : String(cause),
      });
    }
  }

  private resolveVersion(claims: JWTPayload): TokenVersion {
    const ver = claims.ver;
    if (ver === "2.0") return "2.0";
    if (ver === "1.0") return "1.0";
    // Fall back to inferring from the issuer shape when `ver` is absent.
    const iss = typeof claims.iss === "string" ? claims.iss : "";
    if (iss.endsWith("/v2.0")) return "2.0";
    if (iss.includes(this.config.v1IssuerHost)) return "1.0";
    throw new InvalidTokenError("Unable to determine token version", {
      ver: typeof ver === "string" ? ver : null,
    });
  }

  private requireTid(claims: JWTPayload): string {
    const tid = claims.tid;
    if (typeof tid !== "string" || tid.length === 0) {
      throw new InvalidIssuerError("Token is missing the tenant id (tid) claim");
    }
    return tid;
  }

  private assertVersionAccepted(version: TokenVersion): void {
    if (!this.config.acceptedVersions.includes(version)) {
      throw new InvalidTokenError(`Token version ${version} is not accepted`, {
        accepted: this.config.acceptedVersions,
        actual: version,
      });
    }
  }

  private async verifySignatureAndClaims(
    token: string,
    expectedIssuer: string,
  ): Promise<JWTPayload> {
    const verifyOptions = {
      issuer: expectedIssuer,
      audience: [...this.config.audiences],
      algorithms: [...ACCEPTED_ALGORITHMS],
      clockTolerance: this.config.clockToleranceSec,
      currentDate: new Date(this.config.clock.now()),
    };

    try {
      const resolver = await this.jwksClient.getKeyResolver();
      const { payload } = await jwtVerify(token, resolver, verifyOptions);
      return payload;
    } catch (error) {
      // A missing key id usually means the signing keys rotated since we cached
      // them. Force one refresh and retry before giving up.
      if (error instanceof joseErrors.JWKSNoMatchingKey) {
        try {
          const resolver = await this.jwksClient.refresh();
          const { payload } = await jwtVerify(token, resolver, verifyOptions);
          return payload;
        } catch (retryError) {
          throw this.mapJoseError(retryError);
        }
      }
      throw this.mapJoseError(error);
    }
  }

  /** Translate jose / unknown errors into the framework's typed {@link AuthError}s. */
  private mapJoseError(error: unknown): AuthError {
    // Already one of ours (e.g. thrown from the JWKS loader) — pass through.
    if (error instanceof AuthenticationError) return error;

    // NB: JWTExpired extends JWTClaimValidationFailed, so check it first.
    if (error instanceof joseErrors.JWTExpired) {
      return new TokenExpiredError("Token has expired", { claim: "exp" });
    }
    if (error instanceof joseErrors.JWTClaimValidationFailed) {
      switch (error.claim) {
        case "aud":
          return new InvalidAudienceError("Token audience is invalid", {
            expected: this.config.audiences,
          });
        case "iss":
          return new InvalidIssuerError("Token issuer is invalid");
        case "nbf":
          return new TokenNotYetValidError("Token is not yet valid", { claim: "nbf" });
        default:
          return new AuthenticationError(`Token claim '${error.claim}' failed validation`, {
            claim: error.claim,
            reason: error.reason,
          });
      }
    }
    if (error instanceof joseErrors.JWSSignatureVerificationFailed) {
      return new InvalidSignatureError("Token signature verification failed");
    }
    if (error instanceof joseErrors.JWKSNoMatchingKey) {
      return new InvalidSignatureError("No matching signing key was found for the token");
    }
    if (
      error instanceof joseErrors.JWTInvalid ||
      error instanceof joseErrors.JWSInvalid ||
      error instanceof joseErrors.JWTClaimValidationFailed
    ) {
      return new InvalidTokenError("Token is malformed");
    }
    if (error instanceof joseErrors.JOSEError) {
      return new AuthenticationError("Token validation failed", { code: error.code });
    }
    return new AuthenticationError("Unexpected error during token validation");
  }
}

File: packages/go/provider/microsoft/token_validator.go
package microsoft

import (
	"strings"

	"github.com/avi-pathak/easy-sso/packages/go/config"
	"github.com/avi-pathak/easy-sso/packages/go/core"
	"github.com/avi-pathak/easy-sso/packages/go/internal/oidc"
	"github.com/avi-pathak/easy-sso/packages/go/ssoerr"
)

// TokenValidator validates Microsoft Entra ID tokens end-to-end: signature,
// issuer, audience, expiry (exp), not-before (nbf), tenant, and token version.
type TokenValidator struct {
	cfg  config.NormalizedMicrosoftConfig
	jwks *oidc.JWKSClient
}

// NewTokenValidator constructs a TokenValidator.
func NewTokenValidator(cfg config.NormalizedMicrosoftConfig, jwks *oidc.JWKSClient) *TokenValidator {
	return &TokenValidator{cfg: cfg, jwks: jwks}
}

// Validate verifies a token and returns its validated claims. No claim is trusted
// until the signature has been cryptographically verified.
func (v *TokenValidator) Validate(token string) (core.Claims, error) {
	header, claims, parts, err := oidc.Decode(token)
	if err != nil {
		return nil, err
	}

	// Step 1: peek (untrusted) to resolve the expected issuer. With
	// common/organizations the issuer is not fixed — it embeds the real tenant.
	version, err := v.resolveVersion(claims)
	if err != nil {
		return nil, err
	}
	tid, err := requireTid(claims)
	if err != nil {
		return nil, err
	}
	expectedIssuer := ResolveExpectedIssuer(tid, version, v.cfg)

	// Step 2: cryptographic verification + standard claim checks.
	if err := oidc.VerifySignature(v.jwks, parts, header); err != nil {
		return nil, err
	}
	if err := oidc.VerifyStandardClaims(claims, oidc.ClaimsOptions{
		Issuers:   []string{expectedIssuer},
		Audiences: v.cfg.Audiences,
		Now:       v.cfg.Clock.Now(),
		Tolerance: v.cfg.ClockTolerance,
	}); err != nil {
		return nil, err
	}

	// Step 3: policy on the now-verified payload.
	if err := AssertTenantAllowed(tid, v.cfg); err != nil {
		return nil, err
	}
	if err := v.assertVersionAccepted(version); err != nil {
		return nil, err
	}
	return claims, nil
}

func (v *TokenValidator) resolveVersion(claims core.Claims) (config.TokenVersion, error) {
	if ver, ok := claims.String("ver"); ok {
		switch ver {
		case "2.0":
			return config.TokenV2, nil
		case "1.0":
			return config.TokenV1, nil
		}
	}
	// Fall back to inferring from the issuer shape when ver is absent.
	iss, _ := claims.String("iss")
	if strings.HasSuffix(iss, "/v2.0") {
		return config.TokenV2, nil
	}
	if iss != "" && strings.Contains(iss, v.cfg.V1IssuerHost) {
		return config.TokenV1, nil
	}
	return "", ssoerr.NewInvalidTokenError("Unable to determine token version", nil)
}

func requireTid(claims core.Claims) (string, error) {
	if tid, ok := claims.String("tid"); ok {
		return tid, nil
	}
	return "", ssoerr.NewInvalidIssuerError("Token is missing the tenant id (tid) claim", nil)
}

func (v *TokenValidator) assertVersionAccepted(version config.TokenVersion) error {
	for _, a := range v.cfg.AcceptedVersions {
		if a == version {
			return nil
		}
	}
	return ssoerr.NewInvalidTokenError("Token version is not accepted", map[string]any{
		"accepted": v.cfg.AcceptedVersions,
		"actual":   version,
	})
}
